src/main.cpp

```diff
@@ -3296,9 +3325,47 @@ CBlockIndex* AddToBlockIndex(const CBlockHeader& block)
     return pindexNew;
 }
 
+void FallbackSproutValuePoolBalance(
+    CBlockIndex *pindex,
+    const CChainParams& chainparams
+)
+{
+    // We might not want to enable the checkpointing for mainnet
+    // yet.
+    if (!chainparams.ZIP209Enabled()) {
+        return;
+    }
+
+    // Check if the height of this block matches the checkpoint
+    if (pindex->nHeight == chainparams.SproutValuePoolCheckpointHeight()) {
+        if (pindex->GetBlockHash() == chainparams.SproutValuePoolCheckpointBlockHash()) {
+            // Are we monitoring the Sprout pool?
+            if (!pindex->nChainSproutValue) {
+                // Apparently not. Introduce the hardcoded value so we monitor for
+                // this point onwards (assuming the checkpoint is late enough)
+                pindex->nChainSproutValue = chainparams.SproutValuePoolCheckpointBalance();
+            } else {
+                // Apparently we have been. So, we should expect the current
+                // value to match the hardcoded one.
+                assert(*pindex->nChainSproutValue == chainparams.SproutValuePoolCheckpointBalance());
+                // And we should expect non-none for the delta stored in the block index here,
+                // or the checkpoint is too early.
+                assert(pindex->nSproutValue != boost::none);
+            }
+        } else {
+            LogPrintf(
+                "FallbackSproutValuePoolBalance(): fallback block hash is incorrect, we got %s\n",
+                pindex->GetBlockHash().ToString()
+            );
```

Comment:
If we're at the checkpoint height but on the wrong chain, this will fall through, leaving `pindex->nChainSproutValue` as `none`. Then the conditional on line [R2461](https://github.com/zcash/zcash/pull/3885/files#diff-7ec3c68a81efff79b6ca22ac1f1eabbaR2461) will cause the turnstile rule not to be enforced.

It's important that we don't just `assert` that the block hash at the checkpoint height is as expected, because that would introduce a potential DoS during node startup if an attacker created an orphan block at the checkpoint height and sent it to the victim node at just the right time. On the other hand, it seems like we should be able to do better than failing open in this case: in particular if the node reaches 100 blocks past the checkpoint height and is still on the wrong chain, then it's never going to recover. Maybe it's too much complexity to detect that case (if it happens then we have other problems than the turnstile rule).